Library/Homebrew/github_packages.rb

```diff
@@ -38,35 +39,131 @@ def upload_bottles(bottles_hash)
+        }
+        manifest_hash["org.opencontainers.image.description"] = formula.desc if formula.desc.present?
+        manifest_hash["org.opencontainers.image.license"] = formula.license if formula.license.present?
+
+        manifest_annotations = Pathname("#{formula_name}.#{tag}.annotations.json")
+        manifest_annotations.unlink if manifest_annotations.exist?
+        manifest_annotations.write({ "$manifest" => manifest_hash }.to_json)
+
+        # TODO: ideally most/all of these attributes would be stored in the
+        # bottle JSON rather than reading them from the formula.
+        os, arch, os_version = if OS.mac?
+          os = "darwin"
+          macos_version = MacOS::Version.from_symbol(bottle_tag.to_sym)
+          arch = if macos_version.arch == :arm64
+            "arm64"
+          else
+            "amd64"
+          end
+          [os, arch, macos_version.to_f.to_s]
+        elsif OS.linux?
+          ["linux", "amd64", OS.kernel_version.to_s]
+        else
+          raise "Can only upload GitHub Packages on macOS or Linux!"
+        end
+
+        tar_sha256 = Digest::SHA256.hexdigest(
+          Utils.safe_popen_read("gunzip", "--stdout", "--decompress", local_file),
+        )
+
+        config_hash = {
+          "architecture" => arch,
+          "os"           => os,
+          "os.version"   => os_version,
+          "rootfs"       => {
+            "type"     => "layers",
+            "diff_ids" => ["sha256:#{tar_sha256}"],
+          },
+        }
+
+        manifest_config = Pathname("#{formula_name}.#{tag}.config.json")
+        manifest_config.unlink if manifest_config.exist?
+        manifest_config.write(config_hash.to_json)
+
+        image_tag = "#{image}:#{tag}"
+        puts
         system_command!(oras, verbose: true, print_stdout: true, args: [
-          "push", "ghcr.io/#{org}/#{repo}/#{formula_name}:#{tag}",
+          "push", image_tag,
+          "--verbose",
+          "--manifest-annotations=#{manifest_annotations}",
+          "--manifest-config=#{manifest_config}:application/vnd.oci.image.config.v1+json",
           "--username", user,
           "--password", token,
           "#{local_file}:application/vnd.oci.image.layer.v1.tar+gzip"
         ])
 
-        # TODO: make this package public?
-        # TODO: make this latest?
+        image_tag
       end
+
+      image_tag = "#{image}:#{version}#{rebuild}"
+      puts
+      system_command!(docker, verbose: true, print_stdout: true, args: [
+        "buildx", "imagetools", "create", "--tag", image_tag, *image_tags
+      ])
+
+      # TODO: once the main image metadata is working correctly delete the package using:
+      # `curl -X DELETE -u $HOMEBREW_GITHUB_PACKAGES_USER:$HOMEBREW_GITHUB_PACKAGES_TOKEN
+      #  https://api.github.com/orgs/Homebrew/packages/container/homebrew-core%2F$PACKAGE/versions/$VERSION`
+      # Alternatively, remove all usage of `docker` here instead.
```

Comment:
Pushing the architecture specific images to `…:latest` and then referencing them by their SHA-256 instead is an alternative that doesn't require deleting tags.